src/libraries/System.Private.CoreLib/src/System/Type.cs

```diff
@@ -438,8 +438,30 @@ public static Type[] GetTypeArray(object[] args)
             return cls;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static TypeCode GetTypeCode(Type? type)
         {
+            if (RuntimeHelpers.IsKnownConstant(type))
+            {
+                if (type is RuntimeType rt && rt.IsActualEnum)
+                    type = type.GetEnumUnderlyingType();
+                if (type == typeof(sbyte))
+                    return TypeCode.SByte;
+                else if (type == typeof(byte))
+                    return TypeCode.Byte;
+                else if (type == typeof(short))
+                    return TypeCode.Int16;
+                else if (type == typeof(ushort))
+                    return TypeCode.UInt16;
+                else if (type == typeof(int))
+                    return TypeCode.Int32;
+                else if (type == typeof(uint))
+                    return TypeCode.UInt32;
+                else if (type == typeof(long))
+                    return TypeCode.Int64;
+                else if (type == typeof(ulong))
+                    return TypeCode.UInt64;
```

Comment:
This can handle all TypeCode cases now. No reason to limit it to integers.

Reply:
Handling more cases will make the JIT refuse to inline it, @EgorBo said that even the current method will sometimes have trouble with it.

Reply:
This is the common risk with these sorts of patterns. What is the specific JIT limit that would be hit by handling all TypeCode cases here?

FWIW, we have a lot more complex methods in CoreLib that we expect to get inlined with AggressiveInlining. For example, https://github.com/dotnet/runtime/blob/5892ef2a557d83f06eeab9580fee9e2c708244a6/src/libraries/System.Private.CoreLib/src/System/UInt64.cs#L1024-L1095 .

Reply:
> Handling more cases will make the JIT refuse to inline it due to IL limit, @EgorBo said that even the current method will sometimes have trouble with it.

I didn't mean you should maintain some specific limit of cases, just pointed out that in theory it might not be inlined. I assume in that case IsKnownConstant will return false and the whole block will be removed. is that a problem?

Reply:
And even if it'd be fine to add more things here, this is a legacy API that's only useful for enums, handling things that shouldn't be used here faster isn't a good tradeoff for pessimizing inlining.

Reply:
I believe that new code would be better off using `Enum.GetUnderlayingType` and match types instead of using the legacy typecodes.

For existing code that wants to stick with TypeCode, there are patterns where recognizing all TypeCodes is benefical. Take https://github.com/dotnet/runtime/blob/bbcff6bc422da3d9b1904281a86014236ce50f2e/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L2244 for example. If you handle all cases in GetTypeCode, this method should get constant. If you only handle subset of cases, this method is still going to take the slow path in some cases.

We should rewrite this `IsWriteAtomic` method to demonstrate the benefit of the new intrinsics as part of this PR. Ideally, `ConcurrentDictionaryTypeProps` cache would go away as part of that.

Reply:
The issue is with libraries that also target older runtimes, there GetTypeCode will be much faster. Also Mono will too be pessimized by using GetUnderlyingType.

Reply:
That sounds like a point for making `GetTypeCode` work in all cases and not just a subset - so that the libraries that target older runtimes get full benefit when they use patterns like `IsWriteAtomic`.